WIP: add support for Lontium lt6911uxc HDMI-RX to MIPI CSI2-TX

nicolasb565 opened the pull request:
Tested on a rpi cm4 with io-board. I had to remove chip U11 on the io-board.

I don't think the driver support libcamera, and the device tree is hardcoded to use CAM1.

Code based on https://github.com/InES-HPMM/Lontium_lt6911uxc/blob/main/source/lt6911uxc.c
See https://forums.raspberrypi.com/viewtopic.php?t=341781 for discussion.

Commit:
add support for Lontium lt6911uxc HDMI-RX to MIPI CSI2-TX

6by9 commented:
I'll try and review when I get a chance.

Adding any driver is normally 4 patches:
- device tree bindings
- driver
- defconfig changes
- overlay & README.
checkpatch checks that any compatible strings defined in the driver patch are documented in the bindings. Without the binding it will complain.

Only working on CAM1 is fine as I believe it requires 4 data lanes, and that is only present on CAM1. (CAM0 is 2 lanes only)
libcamera is only relevant for raw image sensors, and this device doesn't fall into that category.

Commit:
add lt6911uxc-audio.dts overlay to get audio from lontium lt6911uxc over i2s

Commit:
fix typo in overlays Makefile (.dto extension instead of dtbo for lt6911uxc overlay)

Commit:
use proper filenames for lt6911 overlays

pelwell commented on arch/arm/boot/dts/overlays/Makefile:
```diff
@@ -130,6 +130,8 @@ dtbo-$(CONFIG_ARCH_BCM2835) += \
 	justboom-dac.dtbo \
 	justboom-digi.dtbo \
 	ltc294x.dtbo \
+	lt6911uxc-overlay.dtbo \
```
Don't include the `-overlay` here - there is a make rule to create `*.dtbo` from `*-overlay.dts`.

Also, both new entries belong before `ltc294x.dtbo'.

nicolasb565 replied:
resolved in last commit

pelwell commented on drivers/media/i2c/Makefile:
```diff
@@ -115,6 +115,7 @@ obj-$(CONFIG_VIDEO_S5C73M3)	+= s5c73m3/
 obj-$(CONFIG_VIDEO_ADP1653)	+= adp1653.o
 obj-$(CONFIG_VIDEO_LM3560)	+= lm3560.o
 obj-$(CONFIG_VIDEO_LM3646)	+= lm3646.o
+obj-$(CONFIG_VIDEO_LT6911UXC) += lt6911uxc.o
```
The other lines have a TAB character before the `+=`.

nicolasb565 replied:
resolved in last commit

pelwell reviewed (commented):
Aside from the inline review comments you will need to write README entries for both overlays. Remember to include two blank lines between entries, to use spaces for indents, and to keep to the 80-character line length limit.

Commit:
fix for review

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -2424,6 +2424,21 @@ Info:   This overlay has been deprecated and removed - see gpio-ir
+Info:   Adds support for Lontium lt6911uxc HDMI to CSI
+Load:   dtoverlay=lt6911uxc,<param>=<val>
+Params: media-controller        Configure use of Media Controller API for
+                                configuring the sensor (default off)    
```
Remove the trailing whitespace, please.

nicolasb565 replied:
Resolved in last commit

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -2424,6 +2424,21 @@ Info:   This overlay has been deprecated and removed - see gpio-ir
+        Wiring is LRCK/WFS to GPIO 19, BCK/SCK to GPIO 18, and DATA/SD to GPIO
+        20.
+Load:   dtoverlay=lt6911uxc-audio
+
```
Document the `card-name` parameter.

nicolasb565 replied:
Resolved in last commit

pelwell reviewed (commented):
With those fixed I approve the overlay side of the PR. It still needs @6by9's approval, and dividing into the 4 commits as discussed previously - probably in that order.

6by9 commented on drivers/media/i2c/lt6911uxc_regs.h:
```diff
@@ -0,0 +1,72 @@
+/* SPDX-License-Identifier: (GPL-2.0)
```
Using SPDX licence headers means you don't need the boilerplate of the licence.
SPDX, description, and copyright is sufficient.

Same for the C file(s)

6by9 commented on drivers/media/i2c/lt6911uxc_regs.h:
```diff
@@ -0,0 +1,72 @@
+#define V_SW		0x8671	/* vsync length lines */
+
+#define SYNC_POL	0x8670	/* hsync/vsync polarity flags */
+#define MASK_VSYNC_POL	(1 << 1)
```
Generally use the BIT macro instead of (1 << N) defines.

Commit:
fix for review in overlay README

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+/* custom v4l2 controls */
+#define V4L2_CID_USER_LT6911UXC_BASE (V4L2_CID_USER_BASE + 0x1090)
+#define LT6911UXC_CID_AUDIO_SAMPLING_RATE (V4L2_CID_USER_LT6911UXC_BASE + 1)
+#define LT6911UXC_CID_AUDIO_PRESENT       (V4L2_CID_USER_LT6911UXC_BASE + 2)
```
Ah, the bane of adding custom ioctls - you need to reserve a bank of CID values.
```V4L2_CID_USER_BASE + 0x1090``` clashes with an existing assignment [```V4L2_CID_USER_MAX217X_BASE```](https://elixir.bootlin.com/linux/latest/C/ident/V4L2_CID_USER_MAX217X_BASE).

These really do need to be registered with mainline, otherwise you get conflicts every time they move on. We ran into the same issue with the BCM2835_ISP driver as that is still waiting to be upstreamed #4440 

Regardless, the control block needs to be defined in the uapi headers for use by applications.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+#define POLL_INTERVAL_MS 1000
+
+/* v4l2 dv timings */
+static struct v4l2_dv_timings default_timing = V4L2_DV_BT_CEA_1920X1080P60;
```
This looks odd as it isn't a variable, however the macro for a structure assignment.
This should either be ```const``` as it never changes, or used explicitly in ```lt6911uxc_initial_setup``` (although it should be ```const``` there too).

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+ * Copyright (c) 2020, Lukas Neuner <neur@zhaw.ch>
+ *
+ * This program is based on the tc358840 - Toshiba HDMI to CSI-2 bridge from
+ * Armin Weiss
```
The tc358840 driver isn't in mainline, so this comment is largely irrelevant without a link. 
https://patchwork.linuxtv.org/project/linux-media/patch/44d2d84d-f18d-9c80-977f-8a2b8675e70a@xs4all.nl/ is probably the most useful as Hans Verkuil appears to have taken that driver, tidied it up a fair amount, and at least started upstreaming it.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+	struct i2c_client *i2c_client;
+	struct v4l2_subdev sd;
+	struct media_pad pad;
+	struct mutex lock;
```
What does the mutex protect? It seems to be initialised in ```lt6911uxc_initial_setup```, but then only taken/released in ```lt6911uxc_isr```. I could be wrong, but I don't believe ISRs are re-entrant.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+	mutex_lock(&state->lock);
+	dev_dbg(sd->dev, "%s in kthread %d\n", __func__, current->pid);
+
+	lt6911uxc_ext_control(sd, true);
```
lt6911uxc_ext_control is only called from here, and released below. It therefore doesn't need to keep track of the state in ```enable_i2c```, which drops the complexity.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+		reg, val);
+}
+
+static void lt6911uxc_i2c_rd(struct v4l2_subdev *sd, u16 reg, u8 *values, u32 n)
```
Returning void from an I2C read function which can genuinely fail seems flawed. Logging a message is not error handling, particularly as the values array will not have been touched, so the calling function will quite likely be dealing with uninitialised stack.

nicolasb565 replied:
Yes I agree.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+	switch (sub->type) {
+	case V4L2_EVENT_SOURCE_CHANGE:
+		return v4l2_src_change_event_subdev_subscribe(sd, fh, sub);
+	default:
```
As per https://elixir.free-electrons.com/linux/latest/source/drivers/media/i2c/tc358743.c#L1496, you should allow for subscribing to V4L2_EVENT_CTRL too.

nicolasb565 replied:
What does it do? We don't notify for those type of events?

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+		state->csi_lanes_in_use = lanes;
+		dev_dbg(dev, "MIPI lanes %d\n", lanes);
+
+		lt6911uxc_csi_enable(&state->sd, true);
```
Enabling the CSI interface without checking the streaming status of the device seems a little unusual.

nicolasb565 replied:
What do you mean by that? I don't really know what is considered usual or not.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+
+		lt6911uxc_csi_enable(&state->sd, true);
+
+		if (!state->detected_timings.bt.width) {
```
Why the difference between ```!state->detected_timings.bt.width``` and the normal ```else``` clause. Both update the timings because they've changed, and the different debug message seems an insufficient reason to have a separate clause.

nicolasb565 replied:
Seams like it's to make a distinction between no hdmi input -> hdmi input compared to a simple timing change?

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+			memset(&state->detected_timings, 0,
+			       sizeof(state->detected_timings));
+		}
+		if (handled)
```
This function is only called from one place, and handled is passed in. It seems redundant checking if handled is not NULL.
If the polled thread hadn't passed in a handled variable to lt6911uxc_isr then I would have understood it.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+
+	lt6911uxc_audio_int_handler(state, handled);
+
+	lt6911uxc_log_status(sd);
```
This is going to create a fair amount of logging information if the HDMI input is connected/disconnected frequently. The same info is available via userspace calls, so seems redundant here.

nicolasb565 replied:
Yes I noticed it's quite noisy and I'm going to make it a lot more quiet tomorrow. Only log something when there is an actual change.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+{
+	v4l2_dbg(3, debug, sd, "%s(): enable %d\n", __func__, enable);
+
+	/* handled by ISR */
```
It really shouldn't be. Some CSI2 receivers require an idle bus with LP-11 state when initialising, and you can't achieve that.

The ISR potentially automatically stops on a source change or disconnect, but it shouldn't automatically start.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+	fmt->height = state->timings.bt.height;
+	fmt->field  = V4L2_FIELD_NONE;
+	fmt->colorspace = V4L2_COLORSPACE_SRGB;
+	fmt->ycbcr_enc = V4L2_YCBCR_ENC_601;
+	fmt->quantization = V4L2_QUANTIZATION_LIM_RANGE;
```
Is it really always 601 and limited range? Are the conversion matrices used to get RGB HDMI data to YUYV CSI2 data actually controllable.
What about YCbCr HDMI data which can also be 601, 709, or 2020.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+	case MEDIA_BUS_FMT_UYVY8_1X16:
+		break;
+	default:
+		return -EINVAL;
```
I'm fairly certain this is wrong, but note that TC358743 does the same.
AFAIK V4L2 requires that you return a valid format, not an error code. Same if you get passed in an invalid resolution, it should be fixed up to something that the device can produce.

Admittedly for the video device not subdevice API, but https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-g-fmt.html 
> -EINVAL The struct [v4l2_format](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-g-fmt.html#c.V4L.v4l2_format) type field is invalid or the requested buffer type not supported.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+	if (sd->dev->of_node)
+		return 0;
+
+	/* Support for non-continuous CSI-2 clock is missing in pdate mode */
```
pdata, not pdate.

But we're not using pdata, and that feels very much like a copy/paste rather than necessarily applying here. Confirmation would be nice.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+{
+	struct i2c_client *client = v4l2_get_subdevdata(sd);
+
+	return 0;
```
Why bother? You're doing nothing.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+	.pad	= &lt6911uxc_pad_ops,
+};
+
+#ifdef CONFIG_MEDIA_CONTROLLER
```
Just ```select MEDIA_CONTROLLER``` in Kconfig and remove the ```#ifdef```s. 
You're calling ```media_entity_pads_init``` unconditionally below, so I don't believe this will compile without MEDIA_CONTROLLER being enabled.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+module_i2c_driver(lt6911uxc_driver);
+
+MODULE_DESCRIPTION("Driver for Lontium lt6911uxc HDMI to CSI-2 Bridge");
+MODULE_AUTHOR("Lukas Neuner <neur@zhaw.ch>");
+MODULE_AUTHOR("Alexey Gromov <groo@zhaw.ch>");
```
Do they know that you're listing them as author. Any support requests are likely to head back to them if this is left here.

6by9 commented on drivers/media/i2c/lt6911uxc.c:
```diff
@@ -0,0 +1,989 @@
+// SPDX-License-Identifier: (GPL-2.0)
+/*
+ * lt6911uxc.c - Lontium 4k60 HDMI-CSI bridge driver
```
4k60, but you only allow 4k30. I'd define it as HDMI2.0 or similar rather than a resolution and rate.

6by9 reviewed (commented):
There are a load of ```checkpatch.pl --strict``` errors over alignment. I would like to see this sent upstream, therefore it may as well be clean to start with.

MAINTAINERS really should be updated too. I'd suggest you approach the original authors with regard whether they want to support forks of their code or not.

nicolasb565 commented:
I will address the comment above.

I will add support for actually detecting if the lt6911uxc is there on i2c bus. If the device is not there, I suspect we don't want to create a v4l2 device and we should print some kind of warning?

What I was also thinking was to read the flash and check we got the firmware that was expected. If not, we write the firmware to the flash and restart the Lontium so that it will run properly using the expected firmware.

6by9 commented:
> I will add support for actually detecting if the lt6911uxc is there on i2c bus. If the device is not there, I suspect we don't want to create a v4l2 device and we should print some kind of warning?

If you can check an ID register, then that is a good plan. Most CSI subdevs will, eg https://github.com/torvalds/linux/blob/master/drivers/media/i2c/ov9282.c#L752 called from https://github.com/torvalds/linux/blob/master/drivers/media/i2c/ov9282.c#L1035. It logs an error message and fails the probe if the device isn't present.

> What I was also thinking was to read the flash and check we got the firmware that was expected. If not, we write the firmware to the flash and restart the Lontium so that it will run properly using the expected firmware.

I'd forgotten you'd mentioned that when reviewing the patches. Doing so would be ideal, or initially just checking the firmware version against your supported one, and again erroring out if it differs.

nicolasb565 commented:
> > I will add support for actually detecting if the lt6911uxc is there on i2c bus. If the device is not there, I suspect we don't want to create a v4l2 device and we should print some kind of warning?
> 
> If you can check an ID register, then that is a good plan. Most CSI subdevs will, eg https://github.com/torvalds/linux/blob/master/drivers/media/i2c/ov9282.c#L752 called from https://github.com/torvalds/linux/blob/master/drivers/media/i2c/ov9282.c#L1035. It logs an error message and fails the probe if the device isn't present.
> 
> > What I was also thinking was to read the flash and check we got the firmware that was expected. If not, we write the firmware to the flash and restart the Lontium so that it will run properly using the expected firmware.
> 
> I'd forgotten you'd mentioned that when reviewing the patches. Doing so would be ideal, or initially just checking the firmware version against your supported one, and again erroring out if it differs.

Yes I manage to see if the device is there over i2c. I probe a chip id and check it's the lt6911uxc id.

I'm finishing up the code for firmware programming.

For the firmware blob, how do I include this in the pull request?

6by9 commented:
> Yes I manage to see if the device is there over i2c. I probe a chip id and check it's the lt6911uxc id.

That sounds fine - it's always useful to validate the connected hardware.

> I'm finishing up the code for firmware programming.
>
> For the firmware blob binary, how/where do I include this in the pull request?

Do you have permission to distribute the firmware blob, and under what licence?

For mainline they would normally be submitted via linux-firmware@kernel.org for merging into https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/ alongside the driver.

Seeing as this driver isn't being sent upstream yet, open an issue on https://github.com/RPi-Distro/repo/issues linking to the firmware file and requesting it to be merged into the linux-firmware package, and the relevant person should pick it up.

Commit:
check lt6911uxc device is there on i2c bus in probe

Commit:
lt6911uxc fix remove function, forgot to stop poll timer if no irq line
this would cause our poll handler to get called even if we removed the lt6911uxc overlay

nicolasb565 commented:
> > Yes I manage to see if the device is there over i2c. I probe a chip id and check it's the lt6911uxc id.
> 
> That sounds fine - it's always useful to validate the connected hardware.
> 
> > I'm finishing up the code for firmware programming.
> > For the firmware blob binary, how/where do I include this in the pull request?
> 
> Do you have permission to distribute the firmware blob, and under what licence?
> 
> For mainline they would normally be submitted via [linux-firmware@kernel.org](mailto:linux-firmware@kernel.org) for merging into https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/ alongside the driver.
> 
> Seeing as this driver isn't being sent upstream yet, open an issue on https://github.com/RPi-Distro/repo/issues linking to the firmware file and requesting it to be merged into the linux-firmware package, and the relevant person should pick it up.

We will not do firmware programming in the driver. License and NDA issues... You can buy those chips already programmed by Lontium. They also provide example code to program a new firmware under NDA.

Commit:
notify when we got detected timing change, not only on disconnect

6by9 commented:
I'm not sure where to go with this.

If Lontium have multiple different firmware, then the driver needs to either be written to work with all of them, or verify which version is programmed. Without that there will be numerous support issues that will come back to "wrong firmware, but nothing you can do about that".
That sort of fairly fundamental question needs to be addressed by the mainline Linux folks, not us, therefore the driver will really need to be submitted upstream. I've already given review comments that should head off the initial issues they'd raise, but expect it to be a moderately involved process.

Commit:
make driver more quiet

Commit:
changes for lontium lt6911uxv:
send src change event when audio sampling rate change
work properly with 88k+ sampling rate for audio (would give 0 as sampling rate before)

Commit:
changes for lontium lt6911uxv:
send src change event when audio sampling rate change
work properly with 88k+ sampling rate for audio (would give 0 as sampling rate before)

Commit:
Merge branch 'rpi-5.15.y' of https://github.com/nicolasb565/linux into rpi-5.15.y

Commit:
remove trailing tab

Commit:
add cam0_reg and cam1_reg to CM4 DT

Commit:
remove code duplicate where we send v4l2 event + log twice
